impl/src/error.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright 2020 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt;
+
+pub const PLEASE_FILE_A_BUG: &str =
+  "Please file an issue at github.com/google/cargo-raze with details.";
+
+#[derive(Debug)]
+pub enum RazeError {
+  Generic(String),
+  Internal(String),
+  Rendering {
+    crate_name_opt: Option<String>,
+    message: String,
+  },
+  Planning {
+    dependency_name_opt: Option<String>,
+    message: String,
+  },
+  Config {
+    field_path_opt: Option<String>,
+    message: String,
+  },
+}
+
+impl std::error::Error for RazeError {}
+
+impl fmt::Display for RazeError {
+  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+    match &self {
+      Self::Generic(s) => write!(f, "Raze failed with cause: \"{}\"", s),
+      Self::Internal(s) => write!(
+        f,
+        "Raze failed unexpectedly with cause: \"{}\". {}",
+        s, PLEASE_FILE_A_BUG
+      ),
+      Self::Config {
+        field_path_opt,
+        message,
+      } => match field_path_opt {
```

Comment:
Optional: Can we flatten these matches by moving the subexpression into the match?

```
      Self::Config {
        Some(path),
        message,
      } => {}
      Self::Config {
        None,
        message,
      } => {}
```

?

Reply:
I don't think I understand the request here. Can you elaborate a bit more?

Reply:
Also, if possible, I'd like to prevent making code changes when moving code around so the history is easier to follow. Can I put this into another PR?

Reply:
You're basically asking me not to request changes for your PRs. In general, hard no (come on man, it's a _code review_), but in this case since the code was this way before anyway, OK.

Per the [rust book](https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html), Rust supposedly supports complex match expressions, where the pattern can consist of "nested" elements, meaning that a Type(Field: Optional(Int)) could be matched with Type { Field: Some(5) }. I'm asking you to make this change here, instead of pattern matching into another pattern match.

Reply:
(Will LGTM and merge after ack)

Reply:
I'm definitely **not** asking you to **not** request changes. As I originally asked, I wasn't sure how to implement the changes you had requested. 

I was specifically thinking that because this moves code, that it would be beneficial to keep changes to minimum so if the change were to introduce a regression, that it's easier to find if it were done in a separate PR. I'll read that link ASAP and see if it makes more sense to me

Reply:
I think I understand the changes you were asking for and I think I've correctly implemented them. Is this correct?